core/src/style/textStyle.h

```diff
@@ -34,39 +37,66 @@ class TextStyle : public Style {
         TextLabelProperty::Align align = TextLabelProperty::Align::center;
         LabelProperty::Anchor anchor = LabelProperty::Anchor::center;
 
-        bool isValid() {
-            return fontSize > 0.f && !text.empty();
-        }
+        float fontScale = 1;
+        float lineSpacing = 0;
     };
 
-    TextStyle(std::string _name, std::shared_ptr<FontContext> _fontContext,
-              bool _sdf = false, bool _sdfMultisampling = false,
-              Blending _blendMode = Blending::overlay,
-              GLenum _drawMode = GL_TRIANGLES);
-
-    virtual void onBeginDrawFrame(const View& _view, Scene& _scene) override;
-
-    virtual ~TextStyle();
-
-    bool useSDF() const { return m_sdf; }
-    auto& fontContext() const { return *m_fontContext; }
+    auto& context() const { return m_context; }
 
 protected:
 
-    virtual void constructVertexLayout() override;
-    virtual void constructShaderProgram() override;
+    void constructVertexLayout() override;
+    void constructShaderProgram() override;
 
-    virtual std::unique_ptr<StyleBuilder> createBuilder() const override;
+    std::unique_ptr<StyleBuilder> createBuilder() const override;
 
     bool m_sdf;
-    bool m_sdfMultisampling;
 
-    std::shared_ptr<FontContext> m_fontContext;
+    std::shared_ptr<FontContext> m_context;
 
-    UniformLocation m_uOrtho{"u_ortho"};
-    UniformLocation m_uTex{"u_tex"};
     UniformLocation m_uTexScaleFactor{"u_uv_scale_factor"};
-    
+    UniformLocation m_uTex{"u_tex"};
+    UniformLocation m_uOrtho{"u_ortho"};
+    UniformLocation m_uPass{"u_pass"};
+    UniformLocation m_uMaxStrokeWidth{"u_max_stroke_width"};
+
+    mutable std::vector<std::unique_ptr<DynamicQuadMesh>> m_meshes;
+
+public:
+
+    TextStyle(std::string _name, bool _sdf = false,
+              Blending _blendMode = Blending::overlay,
+              GLenum _drawMode = GL_TRIANGLES);
+
+    /* Create the LabelMeshes associated with FontContext GlyphTexture<s>
+     * No GL involved, called from Tangram::update()
+     */
+    void onBeginUpdate() override;
+
+    /* Upload the buffers of the text batches
+     * Upload the textur atlases
+     */
+    void onBeginFrame() override;
+
+    /* Performs the actual drawing of the meshes in two passes
+     * - First pass if signed distance field is on, draw outlines
+     * - Second pass, draw the inner glyph pixels
+     */
+    void onBeginDrawFrame(const View& _view, Scene& _scene) override;
+
+    DynamicQuadMesh& getMesh(size_t id) const {
+        if (id >= m_meshes.size()) {
```

Comment:
Shouldn't be possible. onBeginUpdate ensures that there are as many meshes as GlyphTextures and FontContext::drawGlyph adds no atlas IDs >= max_textures